radix-engine/src/blueprints/resource/worktop.rs

```diff
@@ -38,24 +39,37 @@ impl WorktopBlueprint {
         api: &mut Y,
     ) -> Result<IndexedScryptoValue, RuntimeError>
     where
-        Y: KernelNodeApi + ClientApi<RuntimeError>,
+        Y: KernelSubstateApi<SystemLockData> + ClientApi<RuntimeError>,
     {
+        // TODO: add `drop` callback for drop atomicity, which will remove the necessity of kernel api.
+
         let input: WorktopDropInput = input.as_typed().map_err(|e| {
             RuntimeError::SystemUpstreamError(SystemUpstreamError::InputDecodeError(e))
         })?;
 
-        let mut node_substates = api.kernel_drop_node(input.worktop.0.as_node_id())?;
-        let substate = node_substates
-            .remove(&OBJECT_BASE_PARTITION)
-            .unwrap()
-            .remove(&WorktopField::Worktop.into())
-            .unwrap();
-        let worktop: WorktopSubstate = substate.as_typed().unwrap();
-        for (_, bucket) in worktop.resources {
+        // Detach buckets from worktop
+        let handle = api.kernel_lock_substate(
```

Comment:
Why not the old style here where we use the substates returned from the dropped node?

Reply:
Ditto, primarily to avoid visiblity loss.